packages/lodestar/src/chain/blocks/post.ts

```diff
@@ -5,77 +5,52 @@ import {IBeaconDb} from "../../db/api";
 import {ILogger} from "@chainsafe/lodestar-utils/lib/logger";
 import {IBeaconMetrics} from "../../metrics";
 import {ChainEventEmitter, IAttestationProcessor} from "../interface";
-import {Epoch} from "@chainsafe/lodestar-types/lib";
+import {ILMDGHOST} from "../forkChoice";
 
 export function postProcess(
   config: IBeaconConfig,
-  db: IBeaconDb,
   logger: ILogger,
+  db: IBeaconDb,
+  forkChoice: ILMDGHOST,
   metrics: IBeaconMetrics,
   eventBus: ChainEventEmitter,
   attestationProcessor: IAttestationProcessor
 ): (source: AsyncIterable<{preState: BeaconState; postState: BeaconState; block: SignedBeaconBlock}>) => Promise<void> {
   return async (source) => {
     return (async function() {
-      for await(const item of source) {
-        await db.processBlockOperations(item.block);
-        await attestationProcessor.receiveBlock(item.block);
-        metrics.currentSlot.set(item.block.message.slot);
-        eventBus.emit("processedBlock", item.block);
-        const preSlot = item.preState.slot;
-        const preFinalizedEpoch = item.preState.finalizedCheckpoint.epoch;
-        const preJustifiedEpoch = item.preState.currentJustifiedCheckpoint.epoch;
-        const currentEpoch = computeEpochAtSlot(config, item.postState.slot);
+      for await(const {block, preState, postState} of source) {
+        await db.processBlockOperations(block);
+        await attestationProcessor.receiveBlock(block);
+        metrics.currentSlot.set(block.message.slot);
+        eventBus.emit("processedBlock", block);
+        const preSlot = preState.slot;
+        const preFinalizedEpoch = preState.finalizedCheckpoint.epoch;
+        const preJustifiedEpoch = preState.currentJustifiedCheckpoint.epoch;
+        const currentEpoch = computeEpochAtSlot(config, postState.slot);
         if (computeEpochAtSlot(config, preSlot) < currentEpoch) {
-          eventBus.emit("processedCheckpoint", {epoch: currentEpoch, root: await db.chain.getBlockRoot(preSlot)});
-          await Promise.all([
-            setJustified(config, db, eventBus, logger, metrics, item.postState, preJustifiedEpoch),
-            setFinalized(config, db, eventBus, logger, metrics, item.postState, preFinalizedEpoch)
-          ]);
-
+          eventBus.emit(
+            "processedCheckpoint",
+            {epoch: currentEpoch, root: forkChoice.getBlockSummaryAtSlot(preSlot).blockRoot},
+          );
+          // newly justified epoch
+          if (preJustifiedEpoch < postState.currentJustifiedCheckpoint.epoch) {
+            logger.important(`Epoch ${postState.currentJustifiedCheckpoint.epoch} is justified!`);
+            metrics.previousJustifiedEpoch.set(preJustifiedEpoch);
+            metrics.currentJustifiedEpoch.set(postState.currentJustifiedCheckpoint.epoch);
+            eventBus.emit("justifiedCheckpoint", postState.currentJustifiedCheckpoint);
+          }
```

Comment:
I liked it more in separate method as it makes everything easier to cover with tests